Animation toolbar styling and UX improvements

leendertvanwolfswinkel opened the pull request:
@reinout @Jonasvsl 
GUI: Add 'difference' checkbox to animation toolbar
GUI: Rename 'infiltration_rate_simple' to 'infiltration rate simple' and 'wet cross section area' to 'wet cross-sectional area'
GUI: Changed the LCD Display to flat for better readability
GUI: Make sure Discharge and Water level are always the first and initially selected items in the parameter combobox
GUI: fit combobox width to contents
GUI: add tooltips
GUI: improved user feedback messages when loading animation layers. This also improves the user feedback messages when loading the model
Styling/layers: improved and data-dependent styling for animation layers
Styling/layers: add color ramps as used in animation layers to QGIS
Styling/layers: Add field z_coordinate (bottom level) to animation node layers
Styling/layers: Add field initial_value to animation layers
Styling/layers: Change name of animation layers according to the selected parameter and mode (difference vs. current)
Styling/layers: Rename layer tree group for animation layers
Styling/layers: Remove animation layers on de-activate
Styling/layers: Remove animation layers group if no layers remain in it (i.e. layers put there by the user accidentally or on purpose)
Code: clearer management of 'enabled' and 'active' states of animation toolbar
Code: in datasource\result_constants.py, added neg_possible to NcVar to keep track of which variables can or can't have negative values (for styling purposes)
Code: Make animation toolbar more specific in which layers to update: no groundwater if model result doesn't have groundwater, only nodes/flowlines if user changes nodes/flowlines combobox value
Code: added docstrings to some functions

leendertvanwolfswinkel committed:
Animation toolbar styling and UX improvements

GUI: Add 'difference' checkbox to animation toolbar
GUI: Rename 'infiltration_rate_simple' to 'infiltration rate simple' and 'wet cross section area' to 'wet cross-sectional area'
GUI: Changed the LCD Display to flat for better readability
GUI: Make sure Discharge and Water level are always the first and initially selected items in the parameter combobox
GUI: fit combobox width to contents
GUI: add tooltips
GUI: improved user feedback messages when loading animation layers. This also improves the user feedback messages when loading the model
Styling/layers: improved and data-dependent styling for animation layers
Styling/layers: add color ramps as used in animation layers to QGIS
Styling/layers: Add field z_coordinate (bottom level) to animation node layers
Styling/layers: Add field initial_value to animation layers
Styling/layers: Change name of animation layers according to the selected parameter and mode (difference vs. current)
Styling/layers: Rename layer tree group for animation layers
Styling/layers: Remove animation layers on de-activate
Styling/layers: Remove animation layers group if no layers remain in it (i.e. layers put there by the user accidentally or on purpose)
Code: clearer management of 'enabled' and 'active' states of animation toolbar
Code: in datasource\result_constants.py, added neg_possible to NcVar to keep track of which variables can or can't have negative values (for styling purposes)
Code: Make animation toolbar more specific in which layers to update: no groundwater if model result doesn't have groundwater, only nodes/flowlines if user changes nodes/flowlines combobox value
Code: added docstrings to some functions

leendertvanwolfswinkel committed:
Update map_animator.py

Bugfix: prevent RuntimeError (C++ object QgsVectorLayer has been deleted) when switching to a new project when animation toolbar is active

leendertvanwolfswinkel committed:
Animation toolbar bugfixes

Style: Correct flowline arrow direction
Style: calculate lines legend class bounds always as if difference checkbox is not checked

leendertvanwolfswinkel committed:
Update map_animator.py

UI: More clarity about what 'relative' mode is and that it relates to nodes only

leendertvanwolfswinkel committed:
Update models.py

UI: More appropriate user feedback

leendertvanwolfswinkel committed:
Fix timeslider enabled state

Bugfix: make timeslider enabled state dependent on active state of sideview tool and map animation tool

leendertvanwolfswinkel committed:
Update threedi_plugin.py

Forgot to update one line in previous commit

leendertvanwolfswinkel committed:
Update map_animator.py

Fix some issues related to activate / deactivate

leendertvanwolfswinkel committed:
Add cells to animation toolbar

for adding the cells layer to gridadmin.sqlite, the output of tool_result_selection.models.DatasourceLayerHelper.get_result_layers() changed.
this method is used in many tools, so all uses of this method have been slightly refactored

leendertvanwolfswinkel committed:
Animation layer sub groups

Seperate 1D and 2D so users can more easily switch between them

leendertvanwolfswinkel committed:
Animation toolbar - Add thresholds styling for velocity

leendertvanwolfswinkel committed:
Update flowline_current.qml

Fix animation toolbar flowline current styling for QGIS < 3.16
Fix animation toolbar flowline current flowline styling for 1d2d links

leendertvanwolfswinkel committed:
Animation Toolbar: fix groundwater issues

Bugfixes for models with groundwater
Calculate legend classes for separately for groundwater and non-groundwater layers
Put groundwater layers in separate group in layer tree

leendertvanwolfswinkel committed:
Update map_animator.py

Fix bug for models without groundwater introduced in previous commit

leendertvanwolfswinkel committed:
Update map_animator.py

Fix bug when removing results of model without groundwater

reinout committed:
Merge branch 'master' into leendert-animation-style

reinout commented:
I've merged master into the branch, this way some of the test setup fixes work. Now I'm at least seeing a test error :-)

reinout commented on datasource/result_constants.py:
```diff
@@ -52,6 +52,8 @@
 # TODO: look at the name
 SUBGRID_MAP_VARIABLES = _Q_TYPES + _H_TYPES  # just take all variables..
 
+NEG_POSSIBLE = dict([(v.name, v.neg_possible) for v in SUBGRID_MAP_VARIABLES])
```
"neg" probably means "negative"? It is better to write it in full.

reinout replied:
"neg" probably means "negative"? It is better to write it in full.

reinout commented on tool_result_selection/models.py:
```diff
@@ -6,9 +6,12 @@
+    get_or_create_flowline_layer,
+    get_or_create_node_layer,
+    get_or_create_pumpline_layer
+)
```
This is a different import style. If you run `make beautiful`, "isort" will change it back to one import per line (as it should be :-) )

reinout commented on tool_sideview/sideview.py:
```diff
@@ -26,6 +26,16 @@ def __init__(self, iface, tdi_root_tool):
         self.widget_nr = 0
+        self._active = False
+
+    @property
+    def active(self):
+        return self._active
+
+    @active.setter
+    def active(self, activate):
+        self._active = activate
+        self.tdi_root_tool.update_slider_enabled_state()
```
Nice and tidy!

reinout commented on utils/color.py:
```diff
@@ -0,0 +1,75 @@
+               'oceanography: Guidelines for effective and accurate colormap selection. Oceanography, 29(3):9-13, '
+               'http://dx.doi.org/10.5670/oceanog.2016.66.'
+     }
+)
```
Very neat.

reinout commented on utils/gridadmin.py:
```diff
@@ -3,14 +3,15 @@
 from osgeo import osr
 from qgis.core import QgsWkbTypes
-from threedigrid.admin.constants import TYPE_FUNC_MAP
+from threedigrid.admin.constants import TYPE_FUNC_MAP, NO_DATA_VALUE
```
Imports on separate lines (=> `make beautiful`)

reinout commented on utils/gridadmin.py:
```diff
@@ -3,14 +3,15 @@
 import numpy as np
 
 
+
```
Extra whiteline. => `black`

reinout commented on utils/gridadmin.py:
```diff
@@ -119,12 +129,27 @@ def save(
+            geom = ogr.Geometry(ogr_wkb_type)
+            if as_cells:
+                if np.all(np.equal(node_data["cell_coords"][:, i], np.array([-9999., -9999., -9999., -9999.]))):
+                    continue
```
I would add a short comment about what this means. "If all cells have a nodata value, don't render anything", whatever.

reinout commented on utils/layer_tree_manager.py:
```diff
@@ -411,7 +411,7 @@ def add_results(self, index, start_row, stop_row):
                     self._mark(group, "result_" + result.file_path.value)
 
-                line, node, pumpline = result.get_result_layers()
+                line, node, cell, pumpline = result.get_result_layers()
```
Perhaps this is not a relevant place to make this comment. And I don't know if anything needs to be done about it. But...:

- In other places it is "pump" instead of "pumpline".
- I've also seen lines, **points**, cells, pumps. so "points" instead of "nodes". I assume that's just some small difference in terminology.

I assume everything is OK, but I'm mentioning it as it *might* indicate a problem somewhere.

leendertvanwolfswinkel replied:
yes it's a mess. but I didn't want to mess with the mess; so I only added 'cell' and left the rest as it is

reinout replied:
Probably a good decision :-)

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+    COLOR_RAMP_OCEAN_DEEP,
+    COLOR_RAMP_OCEAN_HALINE,
+    COLOR_RAMP_OCEAN_CURL
+)
```
`make beautiful` (or `isort .` inside the docker).

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
 STYLES_ROOT = Path(__file__).parent.parent / "layer_styles"
+ANIMATION_LAYERS_NR_LEGEND_CLASSES = 24
+assert ANIMATION_LAYERS_NR_LEGEND_CLASSES % 2 == 0
+
```
Extra newline needed here => `black`

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
 
 STYLES_ROOT = Path(__file__).parent.parent / "layer_styles"
+ANIMATION_LAYERS_NR_LEGEND_CLASSES = 24
+assert ANIMATION_LAYERS_NR_LEGEND_CLASSES % 2 == 0
```
Move the `logger =` line after the `CONSTANTS`.

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+    stops = []
+    for i, color in enumerate(data.colors[1:-1]):
+        stop = QgsGradientStop((i + 1) / (len(data.colors) - 1), QColor(color))
+        stops.append(stop)
```
This looks like the color ramps need a minimum of three colors? Perhaps add a quick comment to the color ramps file. Or, perhaps better, add a quick assert here:

    assert len(data.colors) >= 3, "A color ramp needs at least three colors"

leendertvanwolfswinkel replied:
it needs at least 2 colors, not three. so I added `assert len(data.colors) >= 2` and some logic to deal with 2-value color ramps

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+    ren = lyr.renderer()
+
+    # Set correct legend symbol rotation
+    sym = ren.sourceSymbol().clone()
```
Why not `renderer` and `symbol` instead of the abbreviation?
(Unless it is normal in qgis code: I'm slowly starting to get used to `lyr`...)

leendertvanwolfswinkel replied:
I think it's pretty common from the examples I've seen, but I'll change it to fully written

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+    if variable == 's1':
+        class_attribute_str = str('coalesce(result, z_coordinate)')
+    else:
+        class_attribute_str = str('result')
```
`str('some string')` is the same as `'some string'`, so I don't know why you're turning strings into strings?

leendertvanwolfswinkel replied:
I don't know either, removed it

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+    lyr.renderer().setClassAttribute(class_attribute_str)
+    ren.deleteAllClasses()
+    nr_classes = len(percentiles) - 1
+    percentiles[0] = -9999  # to make nodes / cells are also visible when dry and z_coordinate < percentile[0]
```
Somewhere else I saw `-9999.`, so explicitly floating point. That was in some `np.array` if I remember correctly.

Is this integer value OK?

leendertvanwolfswinkel replied:
I didn't lead to any issues when I tested this, but I changed it anyway (into `float(-9999)`)

reinout commented on utils/styler.py:
```diff
@@ -1,12 +1,193 @@
+        qml_path = STYLES_ROOT / 'tools' / 'animation_toolbar' / 'node_difference.qml'
+    lyr.loadNamedStyle(str(qml_path))
+
+    abs_high = max(abs(percentiles[1]), abs(percentiles[-2]))
+    abs_max = max(abs(percentiles[0]), abs(percentiles[-1]))
```
What is the difference between "high" and "max"?

Perhaps add a quick comment (or change the variable name)?

leendertvanwolfswinkel replied:
added a comment

leendertvanwolfswinkel committed:
black

leendertvanwolfswinkel committed:
isort

leendertvanwolfswinkel committed:
rename neg_possible to negative_possible

reinout commented:
Oh, you can leave the isort/black/`make beautiful` layout items alone. I now see that the rest of the code also needs fixing, so I'll do it myself after merging your branch. Otherwise there are too many changes.

reinout commented:
Note: the travis test error is something that's unrelated to your code. I'm trying to fix that in a separate branch.

But.... do make sure the `docker-compose run qgis-desktop make test` runs for you locally (which it looks like it should).

leendertvanwolfswinkel committed:
several minor changes due to review from Reinout

leendertvanwolfswinkel reviewed (commented):
@reinout thanks for your review, I just made some changes accordingly and pushed them

reinout committed:
Merge branch 'master' into leendert-animation-style

reinout committed:
Adding strace for debugging

reinout committed:
Adjusted pytest-qt to allow for our slightly older pyqt

reinout committed:
Fixed flake8 errors

reinout committed:
Debugging

reinout committed:
Prevent coveralls from failing on forks

reinout committed:
Adding change made by tests to gridadmin.sqlite

My assumption is that there is now an extra 'cell layer' that really should be there.
